spec/app/domain/authentication/authn-azure/validate_azure_annotations_spec.rb

```diff
@@ -0,0 +1,167 @@
+
+    allow(invalid_azure_annotation).to receive(:values)
+                                   .and_return(invalid_azure_annotation)
+    allow(invalid_azure_annotation).to receive(:[])
+                                   .with(:name)
+                                   .and_return("authn-azure/non_existing")
+
+    allow(non_azure_annotation).to receive(:values)
+                                   .and_return(non_azure_annotation)
+    allow(non_azure_annotation).to receive(:[])
+                                   .with(:name)
+                                   .and_return("authn-test/non_existing")
+
+    allow(invalid_azure_annotation_service_id_scoped).to receive(:values)
+                                     .and_return(invalid_azure_annotation_service_id_scoped)
+    allow(invalid_azure_annotation_service_id_scoped).to receive(:[])
+                                                             .with(:name)
+                                                             .and_return("authn-azure/#{good_service_id}/non-existing")
+  end
+
+  #  ____  _   _  ____    ____  ____  ___  ____  ___
+  # (_  _)( )_( )( ___)  (_  _)( ___)/ __)(_  _)/ __)
+  #   )(   ) _ (  )__)     )(   )__) \__ \  )(  \__ \
+  #  (__) (_) (_)(____)   (__) (____)(___/ (__) (___/
+
+  context "A list of annotations" do
+    context "that contains Azure globally-scoped annotations" do
+      context "that are all permitted" do
+        subject do
+          Authentication::AuthnAzure::ValidateAzureAnnotations.new.call(
+            role_annotations: [subscription_id_annotation, resource_group_annotation],
+            service_id: good_service_id
+          )
+        end
+
+        it "does not raise an error" do
+          expect { subject }.to_not raise_error
+        end
+      end
+
+      context "that are not permitted" do
+        subject do
+          Authentication::AuthnAzure::ValidateAzureAnnotations.new.call(
+            role_annotations: [invalid_azure_annotation],
+            service_id: good_service_id
+          )
+        end
+
+        it "raises a ConstraintNotSupported error" do
+          expect { subject }.to raise_error(::Errors::Authentication::AuthnAzure::ConstraintNotSupported)
+        end
+      end
+    end
+
+    context "that contains Azure service-id scoped annotations" do
+      context "that are all permitted" do
+        subject do
+          Authentication::AuthnAzure::ValidateAzureAnnotations.new.call(
+            role_annotations: [subscription_id_annotation_service_id_scoped, resource_group_annotation_service_id_scoped],
+            service_id: good_service_id
+          )
+        end
+
+        it "does not raise an error" do
+          expect { subject }.to_not raise_error
+        end
+      end
+
+      context "that are not permitted" do
+        subject do
+          Authentication::AuthnAzure::ValidateAzureAnnotations.new.call(
+            role_annotations: [invalid_azure_annotation_service_id_scoped],
```

Comment:
I would add an annotation that is permitted to this list. Or maybe have another test for this - one that has only an invalid azure annotation and one that has an invalid azure annotation alongside a valid one.

This can ensure that we don't accept a list of annotations if one of them is valid.